src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/haflows/HaFlowHistorySpec.groovy

```diff
@@ -0,0 +1,258 @@
+import org.openkilda.northbound.dto.v2.haflows.HaFlowPatchPayload
+import org.openkilda.testing.service.northbound.model.HaFlowActionType
+
+import org.springframework.web.client.HttpClientErrorException
+import spock.lang.Narrative
+
+@Narrative("""Verify that history records are created for the basic actions applied to Ha-Flow.""")
+class HaFlowHistorySpec extends HealthCheckSpecification {
+
+    @Tidy
+    def "History records with links details are created during link #type operations and can be retrieved with or without timeline"() {
+        given: "HA-Flow has been created"
+        def swT = topologyHelper.switchTriplets[0]
+        Long timeBeforeOperation = System.currentTimeSeconds()
+        HaFlowExtended haFlow = HaFlowExtended.build(swT, northboundV2, topology).create()
+
+        when: "#type action has been executed"
+        action(haFlow)
+
+        then: "Correct event appears in HA-Flow history and can be retrieved without specifying timeline"
+        def historyRecordWithoutTimeLine = haFlow.getHistory().getEntriesByType(type)
+//        https://github.com/telstra/open-kilda/issues/5320
+        if (type != HaFlowActionType.UPDATE) {
+            assert historyRecordWithoutTimeLine.size() == 1
+        }
+
+        verifyAll {
+            historyRecordWithoutTimeLine[0].haFlowId == haFlow.haFlowId
+            historyRecordWithoutTimeLine[0].taskId
+            historyRecordWithoutTimeLine[0].timestampIso
+
+//            https://github.com/telstra/open-kilda/issues/5366
+//            expectedHistoryRecord[0].payloads.action.find {it == type.getPayloadLastAction()}
+            historyRecordWithoutTimeLine[0].payloads.every { it.timestampIso }
+        }
+
+        and: "Flow history contains all flow properties in the dump section"
+        verifyAll {
+            with(historyRecordWithoutTimeLine.first().dumps[0]) { dump ->
+                dump.dumpType == DumpType.STATE_AFTER
+                dump.haFlowId == haFlow.haFlowId
+
+                dump.maximumBandwidth == haFlow.maximumBandwidth
+                dump.ignoreBandwidth == haFlow.ignoreBandwidth
+                dump.allocateProtectedPath == haFlow.allocateProtectedPath
+                dump.encapsulationType.toString() == FlowEncapsulationType.TRANSIT_VLAN.toString()
+                dump.pathComputationStrategy.toString() == haFlow.pathComputationStrategy.toUpperCase()
+                haFlow.maxLatency ? dump.maxLatency == haFlow.maxLatency * 1000000 : true
+                dump.periodicPings == haFlow.periodicPings
+
+                dump.sharedSwitchId == haFlow.sharedEndpoint.switchId.toString()
+                dump.sharedOuterVlan == haFlow.sharedEndpoint.vlanId
+                dump.sharedInnerVlan == haFlow.sharedEndpoint.innerVlanId
+
+                getSubFlowsEndpoints(dump.haSubFlows).sort() == haFlow.subFlows.endpoint.sort()
+
+            }
+        }
+
+        and: "The same history records can be retrieved with specifying timeline"
```

Comment:
Let's adopt pairwise testing. Instead of retrieving entries `with` and `without` timelines in every test, retrieve entries `with` timeline in some test and `without` timeline in others. E.g. `records WITH link can be retrieved WITHOUT timeline` + `records WITHOUS link can be retrieved WITH timeline`.

Reply:
sure.